src/main/cpp/src/timezones.cu

```diff
@@ -118,6 +121,144 @@ auto convert_timestamp_tz(column_view const& input,
+    }
+  }
+};
+
+std::unique_ptr<column> convert_to_utc_with_multiple_timezones(
+  column_view const& input_seconds,
+  column_view const& input_microseconds,
+  column_view const& invalid,
+  column_view const& tz_type,
+  column_view const& tz_offset,
+  table_view const& transitions,
+  column_view const tz_indices,
+  rmm::cuda_stream_view stream,
+  rmm::device_async_resource_ref mr)
+{
+  CUDF_EXPECTS(input_seconds.type().id() == cudf::type_id::INT64,
+               "seconds column must be of type INT64");
+  CUDF_EXPECTS(input_microseconds.type().id() == cudf::type_id::INT32,
+               "microseconds column must be of type INT32");
+
+  // get the fixed transitions
+  auto const ft_cdv_ptr        = column_device_view::create(transitions.column(0), stream);
+  auto const fixed_transitions = lists_column_device_view{*ft_cdv_ptr};
+
+  auto result = cudf::make_timestamp_column(cudf::data_type{cudf::type_to_id<cudf::timestamp_us>()},
+                                            input_seconds.size(),
+                                            rmm::device_buffer{},
+                                            0,
+                                            stream,
+                                            mr);
+  auto null_mask = cudf::make_fixed_width_column(cudf::data_type{cudf::type_id::UINT8},
+                                                 input_seconds.size(),
+                                                 cudf::mask_state::UNALLOCATED,
+                                                 stream,
+                                                 mr);
+
+  thrust::for_each_n(rmm::exec_policy_nosync(stream),
+                     thrust::make_counting_iterator<size_type>(0),
+                     input_seconds.size(),
+                     convert_with_timezones_fn{input_seconds.begin<int64_t>(),
+                                               input_microseconds.begin<int32_t>(),
+                                               invalid.begin<uint8_t>(),
+                                               tz_type.begin<uint8_t>(),
+                                               tz_offset.begin<int32_t>(),
+                                               fixed_transitions,
+                                               tz_indices.begin<int32_t>(),
+                                               result->mutable_view().begin<cudf::timestamp_us>(),
+                                               null_mask->mutable_view().begin<uint8_t>()});
+
+  auto [output_bitmask, null_count] = cudf::detail::valid_if(null_mask->view().begin<uint8_t>(),
+                                                             null_mask->view().end<uint8_t>(),
+                                                             cuda::std::identity{},
+                                                             stream,
+                                                             mr);
+  result->set_null_mask(std::move(output_bitmask), null_count);
```

Comment:
Typically we should avoid keeping the null mask if there is no nulls. Some downstream functions may take extra time to process that null mask without doing any meaningful work.
```suggestion
  if (null_count) { result->set_null_mask(std::move(output_bitmask), null_count); }
```